### PR TITLE
Make RateLimiter::GetTotalPendingRequest() non pure virtual for backward compability

### PR DESCRIPTION
**Pull request by hx235**
Context/Summary:
https://github.com/facebook/rocksdb/pull/8890 added a public API `RateLimiter::GetTotalPendingRequest()` but mistakenly marked it as pure virtual, forcing RateLimiter's derived classes to implement this function and breaking backward compatibility.

This PR makes `RateLimiter::GetTotalPendingRequest()` as non-pure virtual method by providing a trivial implementation in rate_limiter.h

Test:
Passing existing tests

### CODE REVIEW

**Review by pdillinger — CHANGES_REQUESTED** [review]
This line in the HISTORY doesn't fully make sense anymore:

HISTORY.md:* Add a public API RateLimiter::GetTotalPendingRequests() for the total number of requests that are pending for bytes in the rate limiter.

It appears this function is not used by RocksDB internally but is provided by RocksDB built-in rate limiter implementations for use by user code.

And relevant to that: I suggest making the function return a more distinctive "not implemented" value such as -1 and also "assert(false)" for misuse detection in debug builds.

**Review by pdillinger — COMMENTED** [review]
Actually, I think the most "accurate" way to add this feature is to not have this function in RateLimiter but put it in BuiltinRateLimiter derived from RateLimiter, and have NewGenericRateLimiter return BuiltinRateLimiter. But I'm not sure it's worth the extra complexity to be "accurate" with our types. (Other opinions welcome.)

**Comment by hx235** [comment]
Update:
- Updated HISTORY.md/trivial implementation of `RateLimiter::GetTotalPendingRequests()`  as suggested

**Comment by ajkr** [comment]
I am guessing it is reasonable for a non-builtin `RateLimiter` (such as user wrappers, which we may see soon) to implement this function.

I thought a more traditional way to implement this would be to return a `Status::NotSupported()` by default, and overriders can assign the result to a pointer argument and return `Status::OK()`. Then callers can write code that uses the result when available or does something else when it's not supported in their `RateLimiter` object.

**Comment by pdillinger** [comment]
It could be called ExtendedRateLimiter or RateLimiterV2 or something.

I'm not a big fan of Status for things that aren't subject to external factors, but ok in a case like this.

File: include/rocksdb/rate_limiter.h [hunk 1]
```diff
@@ -91,7 +91,11 @@ class RateLimiter {
 
   // Total # of requests that are pending for bytes in rate limiter
```
Comment (pdillinger) [inline include/rocksdb/rate_limiter.h:92]: We need more description in here like

    For convenience, this function is implemented by the RateLimiter returned by NewGenericRateLimiter but is not required by RocksDB. The default implementation indicates "not supported."

**Review by pdillinger — APPROVED** [review]
Otherwise LGTM

**Review by pdillinger — COMMENTED** [review]
Ideally, we should also have a smoke test with a custom RateLimiter (in test code) not implementing GetTotalPendingRequests, to hopefully detect if we start calling the function.

**Comment by hx235** [comment]
Update:
- Added comment for `RateLimiter::GetTotalPendingRequests()` as suggested
- Added a smoke test to verify the optional `RateLimiter::GetTotalPendingRequests()` is not required for basic RocksDB operation

**Comment by hx235** [comment]
Note: haven't imported the latest PR to fbsource since there is an error "An error occurred while syncing the pull request data from GitHub; please try again in a few minutes."

**Comment by pdillinger** [comment]
LGTM :)

**Comment by hx235** [comment]
Update
- Removed unnecessary semicolon after member function definition
- Added key word "override" for overriding virtual destructor

File: db/db_test.cc [hunk 2]
```diff
@@ -3937,6 +3937,56 @@ TEST_F(DBTest, DISABLED_RateLimitingTest) {
...
+  int64_t GetTotalRequests(
+      const Env::IOPriority pri = Env::IO_TOTAL) const override {
+    (void)pri;
+    return 0;
+  }
+
+  int64_t GetBytesPerSecond() const override { return 0; }
+};
+
+// To test that customed rate limiter not implementing optional APIs (e.g,
```
Comment (ajkr) [inline db/db_test.cc:3977]: Honestly I find this solution strange (now there's an extension API that we're not supposed to call? Seems unnecessarily restrictive.). I hope in the future we can just follow the convention of returning Status::NotSupported.

Comment (hx235) [reply db/db_test.cc:3977]: Discussed offline

**Pull request closed without merging**